zarr/hierarchy.py

```diff
@@ -0,0 +1,269 @@
+        return Group(store, readonly=self.readonly, name=names[-1])
+
+    def require_group(self, name):
+
+        # handle compound request
+        names = [s for s in name.split('/') if s]
+        if not names:
+            raise ValueError(name)
+
+        # create groups
+        store = self.store
+        for name in names:
+            store = store.require_store(name)
+            if check_array(store):
+                raise KeyError(name)  # TODO better error?
+            elif check_group(store):
+                pass
+            else:
+                init_group(store)
+
+        return Group(store, readonly=self.readonly, name=names[-1])
+
+    def create_dataset(self, name, data=None, shape=None, chunks=None,
+                       dtype=None, compression='default',
+                       compression_opts=None, fill_value=None, order='C',
+                       synchronizer=None, **kwargs):
+
+        # handle compound request
+        names = [s for s in name.split('/') if s]
+        if not names:
+            raise ValueError(name)
+
+        # create intermediate groups
+        store = self.store
+        for name in names[:-1]:
+            store = store.require_store(name)
+            if check_array(store):
+                raise KeyError(name)  # TODO better error?
+            elif check_group(store):
+                pass
+            else:
+                init_group(store)
+
+        # create store to hold the new array (must not exist)
+        store = store.create_store(names[-1])
+
+        # compatibility with h5py
+        fill_value = kwargs.get('fillvalue', fill_value)
+
+        if name in self:
+            raise KeyError(name)
+
+        if data is not None:
+            a = array(data, chunks=chunks, dtype=dtype,
+                      compression=compression,
+                      compression_opts=compression_opts,
+                      fill_value=fill_value, order=order,
+                      synchronizer=synchronizer, store=store)
+
+        else:
+            a = create(shape=shape, chunks=chunks, dtype=dtype,
+                       compression=compression,
+                       compression_opts=compression_opts,
+                       fill_value=fill_value, order=order,
+                       synchronizer=synchronizer, store=store)
+
+        return a
+
+    def require_dataset(self, name, shape=None, dtype=None,
+                        exact=False, **kwargs):
+        # TODO
+        pass
+
+
+def group(store=None, readonly=False):
+    """TODO"""
+    if store is None:
+        store = MemoryStore(readonly=readonly)
+    init_group(store)
+    return Group(store, readonly=readonly)
+
+
+def open_group(path, mode='a'):
+    """TODO"""
+
+    # ensure directory exists
+    if not os.path.exists(path):
+        if mode in ['w', 'w-', 'x', 'a']:
+            os.makedirs(path)
+        elif mode in ['r', 'r+']:
+            raise ValueError('path does not exist: %r' % path)
```

Comment:
So `open_group` is intended as just a convenience function for people commonly working with data on disk. Similar to how `open` is also a convenience function for working with an array on disk.

The general API is to instantiate a `store` then pass that into the constructor for the `Group` class.